### ISSUE TITLE
Static Method annotated with @DataTableType requires instance of class containing it in order to be invoked, contrary to Reflection requirements.

### ISSUE THREAD

**Issue by BuzMack** [cucumber/cucumber-jvm#1950]
**Describe the bug**
When registering a DataTableType as a static method within the class which represents that Type, then when transforming a datatable to that type the given class is attempted to be instantiated by dependency injection (which is unneeded) which fails.

From the following scenarios:
```gherkin
  Scenario: Create TestBeans from a datatable transform
    Given generate TestBeans from
      | ItemOne | ItemTwo |
      | first   | second  |
      | alpha   | beta    |

  Scenario: Create TestBean directly
    Given generate TestBean from "foo" and "bar"
```

The second scenario yields this expected output:
```
Scenario: Create TestBean directly             # src/feature/TestBeans.feature:10
generated single TestBean: TestBean{stringOne='foo', stringTwo='bar'}
  Given generate TestBean from "foo" and "bar" # xxxxx.testlibrary.TestStepLibrary.generateTestBeans(java.lang.String,java.lang.String)
```

The first scenario fails:
```
Scenario: Create TestBeans from a datatable transform # src/feature/TestBeans.feature:4
  Given generate TestBeans from                       # xxxxx.testlibrary.TestStepLibrary.generateTestBeans(io.cucumber.datatable.DataTable)
      io.cucumber.datatable.CucumberDataTableException: 'java.util.List<xxxxx.testlibrary.TestBean>' could not transform
      | ItemOne | ItemTwo |
      | first   | second  |
      | alpha   | beta    |

	at io.cucumber.datatable.DataTableType.transform(DataTableType.java:144)
	at io.cucumber.datatable.DataTableTypeRegistryTableConverter.toListOrProblems(DataTableTypeRegistryTableConverter.java:151)
	at io.cucumber.datatable.DataTableTypeRegistryTableConverter.toList(DataTableTypeRegistryTableConverter.java:131)
	at io.cucumber.datatable.DataTable.asList(DataTable.java:181)
	at xxxxx.testlibrary.TestStepLibrary.generateTestBeans(TestStepLibrary.java:43)
	at ✽.generate TestBeans from(file:///E:/DevWorkspace/CucumberDataTableTransform/src/feature/TestBeans.feature:5)
Caused by: org.picocontainer.PicoCompositionException: 3 satisfiable constructors is too many for 'class java.lang.String'. Constructor List:[<init>(<init>Buffer), <init>(<init>Builder), <init>()]
	at org.picocontainer.injectors.ConstructorInjector.getGreediestSatisfiableConstructor(ConstructorInjector.java:189)
```

In the above scenarios, the TestBean class is defined as follows:
```java
package xxxxx.testlibrary;

import io.cucumber.java.DataTableType;

import java.util.Map;

public class TestBean {
    private String stringOne;
    private String stringTwo;

    public TestBean(String stringOne, String stringTwo) {
        this.stringOne = stringOne;
        this.stringTwo = stringTwo;
    }

    public String getStringOne() {
        return stringOne;
    }

    public String getStringTwo() {
        return stringTwo;
    }

    @Override
    public String toString() {
        return "TestBean{" +
                "stringOne='" + stringOne + '\'' +
                ", stringTwo='" + stringTwo + '\'' +
                '}';
    }

    @DataTableType
    public static TestBean makeTestBean(Map<String, String> row) {
        return new TestBean(
                row.get("ItemOne"),
                row.get("ItemTwo")
        );
    }
}
```

**To Reproduce**
Implement the two steps defined in the example above:
```java
package xxxxx.testlibrary;

import io.cucumber.datatable.DataTable;
import io.cucumber.java.After;
import io.cucumber.java.Before;
import io.cucumber.java.en.Given;
import io.cucumber.java.Scenario;

import java.util.List;

public class TestStepLibrary {
    protected Scenario scenario;

    @Before
    public void doBefore(Scenario scenario) {
        this.scenario = scenario;
    }

    @After
    public void doAfter(Scenario scenario) {

    }

    @Given("^generate TestBeans from$")
    public void generateTestBeans(DataTable dataTable) {
        List<TestBean> testBeans = dataTable.asList(TestBean.class);
        int count = 0;
        for(TestBean testBean : testBeans) {
            System.out.format("generated TestBean %s: %s%n", count++, testBean);
        }
    }

    @Given("^generate TestBean from \"([^\"]+)\" and \"([^\"]+)\"$")
    public void generateTestBeans(String item1, String item2) {
        TestBean testBean = new TestBean(item1, item2);
        System.out.format("generated single TestBean: %s%n", testBean);
    }
}
```

**Expected behavior**
The dataTable transform to TestBean object behaves as though the @DataTableType annotation and method were provided in a separate class, in this case the transform functions without needing to instantiate the TestBean class since the transform method is defined as a public static method.

**Context & Motivation**
We have multiple Cucumber step libraries which are pulled into various tests depending on testers' needs; most of these libraries are optional. Additionally, we also have many developers working on each library, and the simplest (and thus far easiest) guidance we've been able to provide with using dataTable transforms has been to keep both the transformed type and the method which defines that transformed type within the same class, and to ensure that the method is defined as a ```public static``` since Reflection does not require the containing class to be instantiated when calling a static method.

To facilitate that recommendation (for Cucumber 4), we have a custom method annotation which we scan for prior to launching tests (where the methods themselves are public statics, and have zero parameters), each of the methods found are added to the typeRegistry.
As a simple example, this bean class:
```java
public class DepositFromVar {
    private String toName;
    private String fromValue;

    public DepositFromVar(String toName, String fromValue) {
        this.toName = toName;
        this.fromValue = fromValue;
    }

    public String getToName() {
        return toName;
    }

    public String getFromValue() {
        return fromValue;
    }

    @CustomDataTableTypeDef
    public static DataTableType DefineDataTableType() {
        return new DataTableType(
                DepositFromVar.class,
                (Map<String, String> row) -> new DepositFromVar(
                        row.get("toName"),
                        row.get("fromValue")
                )
        );
    }
}
```

With the above bean class, the ```DefineDataTableType()``` method would be discovered by our scanner and then added to the typeRegistry (in very simplified form here):
```java
for (Method DefineDataTableTypeMethod : dataTableTypeDefMethods) {
    // Method exists, has the correct return type and parameter count
    try {
        DataTableType newDataTableType = (DataTableType) DefineDataTableTypeMethod.invoke(null);
        typeRegistry.defineDataTableType(newDataTableType);
    }
}
```

With the changes introduced in Cucumber 5, our custom scanning component is basically no longer required, however the expectation that the static method can be invoked without instantiating the class in which it is contained becomes problematic in this particular case.

This bug is not a deal-breaker or absolute blocker however since the @DataTableType annotation and attached method can still exist in a separate class, it just requires more thought and communications between our developers than they're currently used to.

**Your Environment**
- Versions used:
  - jackson-databind 2.9.8
  - cucumber libraries 4.4.0 and 5.6.0
  - junit 4.12

**Comment by mpkorstanje** [comment]
Interesting find. This is relatively easy to fix but may break expectations the other way around so I'd rather fix this on the next major release.

**Comment by BuzMack** [comment]
> Interesting find. This is relatively easy to fix but may break expectations the other way around so I'd rather fix this on the next major release.

Fair enough. As mentioned, we already have a workaround which functions as expected, so it's not a blocker... just surprising. Thanks!

**Comment by mpkorstanje** [comment]
Fixed by: 97c1162
